### PR TITLE
[CI][Bugfix] Fix the error in generating the performance data table and add a fallback mechanism that prevents the result file from being generated when test case execution fails.

### LINKED ISSUE #2838: [CI Failuer]: nightly ci,  Nightly Collection & Email, [2026-04-15T19:55:44Z] ValueError: Cannot convert {'mean_ttft_ms': 1000, 'mean_audio_ttfp_ms': 30000, 'mean_audio_rtf': 0.35} to Excel
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
ci env
```

</details>

### Your code version

<details>
<summary>The commit id or version of vllm</summary>

```text
0.19.0
```
</details>
<details>
<summary>The commit id or version of vllm-omni</summary>

```text
e9581137e9d887c0876885d1c4a74ea7d63ba2eb
```
</details>

### 🐛 Describe the bug

https://buildkite.com/vllm/vllm-omni/builds/6825/steps/canvas?sid=019d9257-25f8-4571-8613-711dd44945e4&tab=output
```
[2026-04-15T19:55:44Z] Traceback (most recent call last):
--
[2026-04-15T19:55:44Z]   File "/var/lib/buildkite-agent/builds/bk-cpu-queue-premerge-i-0a0f2e803c3741bff-1/vllm/vllm-omni/tools/nightly/generate_nightly_perf_excel.py", line 856, in <module>
[2026-04-15T19:55:44Z]     main()
[2026-04-15T19:55:44Z]   File "/var/lib/buildkite-agent/builds/bk-cpu-queue-premerge-i-0a0f2e803c3741bff-1/vllm/vllm-omni/tools/nightly/generate_nightly_perf_excel.py", line 845, in main
[2026-04-15T19:55:44Z]     generate_excel_report(
[2026-04-15T19:55:44Z]   File "/var/lib/buildkite-agent/builds/bk-cpu-queue-premerge-i-0a0f2e803c3741bff-1/vllm/vllm-omni/tools/nightly/generate_nightly_perf_excel.py", line 819, in generate_excel_report
[2026-04-15T19:55:44Z]     _write_sheet(ws_omni_raw, omni_raw_columns, omni_sorted)
[2026-04-15T19:55:44Z]   File "/var/lib/buildkite-agent/builds/bk-cpu-queue-premerge-i-0a0f2e803c3741bff-1/vllm/vllm-omni/tools/nightly/generate_nightly_perf_excel.py", line 684, in _write_sheet
[2026-04-15T19:55:44Z]     ws.append(row_values)
[2026-04-15T19:55:44Z]   File "/var/lib/buildkite-agent/.local/lib/python3.9/site-packages/openpyxl/worksheet/worksheet.py", line 673, in append
[2026-04-15T19:55:44Z]     cell = Cell(self, row=row_idx, column=col_idx, value=content)
[2026-04-15T19:55:44Z]   File "/var/lib/buildkite-agent/.local/lib/python3.9/site-packages/openpyxl/cell/cell.py", line 119, in __init__
[2026-04-15T19:55:44Z]     self.value = value
[2026-04-15T19:55:44Z]   File "/var/lib/buildkite-agent/.local/lib/python3.9/site-packages/openpyxl/cell/cell.py", line 218, in value
[2026-04-15T19:55:44Z]     self._bind_value(value)
[2026-04-15T19:55:44Z]   File "/var/lib/buildkite-agent/.local/lib/python3.9/site-packages/openpyxl/cell/cell.py", line 187, in _bind_value
[2026-04-15T19:55:44Z]     raise ValueError("Cannot convert {0!r} to Excel".format(value))
[2026-04-15T19:55:44Z] ValueError: Cannot convert {'mean_ttft_ms': 1000, 'mean_audio_ttfp_ms': 30000, 'mean_audio_rtf': 0.35} to Excel
[2026-04-15T19:55:44Z] 🚨 Error: The command exited with status 1
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by yenuo26**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
**Fix https://github.com/vllm-project/vllm-omni/issues/2838: a nightly CI bug in performance report generation**

- Resolve the Excel generation failure caused by non-Excel-compatible structured values in benchmark result fields (e.g., nested dict objects), which previously broke the Nightly Collection & Email stage.
- Add corresponding unit test cases.

**Improve performance dashboard data continuity with a fallback mechanism**
Introduce a degradation path for benchmark pipelines so that when perf result files are not generated as expected (e.g., benchmark execution exceptions), template-based fallback results are produced and uploaded to the same expected output location/filename pattern.
This ensures downstream report generation (Excel/HTML) and dashboard ingestion can continue with consistent artifacts instead of being completely interrupted.

**Restore the accuracy test that was lost during the previous merge.**

## Test Plan
1.run new ut test in local
```
pytest -sv tests/test_generate_nightly_perf_html.py
pytest -sv tests/test_generate_nightly_perf_excel.py
```
2.Use real performance results to locally simulate the generation of performance Excel and HTML files.
```
python tools/nightly/generate_nightly_perf_html.py
python tools/nightly/generate_nightly_perf_excel.py
```
3.Simulate fallback handling for exceptional cases locally.
4.run perf test in ci

## Test Result
1.
a2f2540b027b6272e34473661556fce3
430b93a77a98cc8375baea5f57ec0487

2.
[nightly_perf_20260416-033505.xlsx](https://github.com/user-attachments/files/26774315/nightly_perf_20260416-033505.xlsx)
[nightly_perf_20260416-033536_v2.html](https://github.com/user-attachments/files/26774319/nightly_perf_20260416-033536_v2.html)

3.
**run omni benchmark**
ab3bfae4e14ab7561ba00b74a740312a
```
{
  "date": "",
  "endpoint_type": "",
  "backend": "",
  "label": null,
  "model_id": "",
  "tokenizer_id": "",
  "num_prompts": 0,
  "request_rate": "",
  "burstiness": 0,
  "max_concurrency": 0,
  "duration": 0,
  "completed": 0,
  "failed": 0,
  "total_input_tokens": 0,
  "total_output_tokens": 0,
  "request_throughput": 0,
  "request_goodput": null,
  "output_throughput": 0,
  "total_token_throughput": 0,
  "total_audio_duration_s": 0,
  "total_audio_frames": 0,
  "audio_throughput": 0,
  "max_output_tokens_per_s": 0,
  "max_concurrent_requests": 0,
  "rtfx": 0,
  "mean_ttft_ms": 0,
  "median_ttft_ms": 0,
  "p99_ttft_ms": 0,
  "mean_tpot_ms": 0,
  "median_tpot_ms": 0,
  "p99_tpot_ms": 0,
  "mean_itl_ms": 0,
  "median_itl_ms": 0,
  "p99_itl_ms": 0,
  "mean_e2el_ms": 0,
  "median_e2el_ms": 0,
  "p99_e2el_ms": 0,
  "mean_audio_rtf": 0,
  "median_audio_rtf": 0,
  "p99_audio_rtf": 0,
  "mean_audio_ttfp_ms": 0,
  "median_audio_ttfp_ms": 0,
  "p99_audio_ttfp_ms": 0,
  "mean_audio_duration_s": 0,
  "median_audio_duration_s": 0,
  "p99_audio_duration_s": 0,
  "baseline": {
    "mean_ttft_ms": 2000,
    "mean_audio_ttfp_ms": 2000,
    "mean_audio_rtf": 0.25
  },
  "random_input_len": 100,
  "random_output_len": 100
}
```

**run diffusion benchmark**
67e6eeb6aa5044d387a4e63e6a1cbf06
```
[
  {
    "test_name": "test_qwen_image_single_device",
    "backend": "vllm-omni",
    "timestamp": "20260416-064044",
    "server_params": {
      "model": "Qwen/Qwen-Image",
      "serve_args": {
        "enable-diffusion-pipeline-profiler": true
      }
    },
    "benchmark_params": {
      "name": "512x512_steps20",
      "dataset": "random",
      "task": "t2i",
      "width": 512,
      "height": 512,
      "num-inference-steps": 20,
      "num-prompts": 10,
      "max-concurrency": 1,
      "enable-negative-prompt": true,
      "baseline": {
        "throughput_qps": 0.3,
        "latency_mean": 3.5,
        "peak_memory_mb_mean": 67000
      }
    },
    "result": {
      "duration": 0,
      "completed_requests": 0,
      "failed_requests": 0,
      "throughput_qps": 0,
      "latency_mean": 0,
      "latency_median": 0,
      "latency_p99": 0,
      "latency_p95": 0,
      "latency_p50": 0,
      "peak_memory_mb_max": 0,
      "peak_memory_mb_mean": 0,
      "peak_memory_mb_median": 0,
      "stage_durations_mean": {},
      "stage_durations_p50": {},
      "stage_durations_p99": {},
      "backend": "",
      "model": "",
      "dataset": "",
      "task": ""
    },
    "log_file": "/app/vllm-omni-yn/tests/dfx/perf/results/logs/test_qwen_image_single_device_vllm-omni_20260416-064044.log",
    "Model": "Qwen/Qwen-Image",
    "Framework": "vllm-omni",
    "Hardware": "",
    "Deployment": "",
    "Task": "t2i",
    "Dataset": "random",
    "resolution": "512x512",
    "Parallelism": "none",
    "max_concurrency": 1,
    "Cache": "disabled",
    "Quantization": "disabled",
    "offload": "disabled",
    "compile": "enabled",
    "Attn_backend": "FLASH_ATTN",
    "num_inference_steps": 20,
    "completed": 0,
    "failed": 0,
    "throughput_qps": 0,
    "latency_mean": 0,
    "latency_median": 0,
    "latency_p99": 0,
    "latency_p95": 0,
    "latency_p50": 0,
    "peak_memory_mb_max": 0,
    "peak_memory_mb_mean": 0,
    "peak_memory_mb_median": 0,
    "stage_durations_mean": {},
    "stage_durations_p50": {},
    "stage_durations_p99": {},
    "commit_sha": "58efc5dde3d27f9912d1265b010294a5e53ea026",
    "build_id": "",
    "build_url": "",
    "source_file": "tests/dfx/perf/tests/test_qwen_image_vllm_omni.json"
  }
]
```

4.
omni perf test
80d00c0a-e62b-49fb-b039-9abffdecc5de

diffusion perf and acc test
2489d11f-bd26-4e69-8172-8f006db792bc

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by yenuo26** [comment]
@amy-why-3459 @fhfuih PTAL

**Comment by hsliuustc0106** [comment]
## Blocking Issues

1. **[Reliability/Safety] .buildkite/test-nightly.yml** - Using `set +e` and `|| true` for artifact upload means benchmark failures no longer cause immediate pipeline failure. This could mask actual test failures downstream since the exit code is only propagated at the very end. Consider whether the fallback mechanism should also trigger CI failure alerts.

2. **[Reliability/Safety] tests/dfx/perf/scripts/run_benchmark.py:69** - Stage init timeout increased from 300s to 600s without justification in the PR description or comments. Is there a specific reason for this increase? Large timeouts can delay failure detection.

---

**VERDICT: REQUEST_CHANGES**

The `set +e` pattern with delayed exit code propagation makes it harder to correlate CI failures with specific benchmark steps. Consider using a wrapper script that generates the fallback template AND records the original failure reason for observability.

Please document why `--stage-init-timeout` was increased from 300 to 600 seconds. If this is to accommodate slower models, consider making it configurable per-test.

The template files (`result_omni_template.json`, `diffusion_result_template.json`) use empty strings `"`" for many fields. Consider using `null` for missing values to make it clear in downstream reports that the data is missing (not just an empty value).

**Comment by yenuo26** [comment]
> Please document why `--stage-init-timeout` was increased from 300 to 600 seconds. If this is to accommodate slower models, consider making it configurable per-test.

1. Using set +e is only to prevent Buildkite from exiting immediately when pytest reports an error, which would otherwise cause subsequent upload operations to be ignored. The relevant error messages will still be displayed normally in the Buildkite logs.

2. Due to the increased overall runtime caused by serial execution, the temporary workaround before this issue is fixed is to extend the timeout. In fact, a unified adjustment had already been made previously; this parameter was mistakenly changed to 300 in an earlier PR, and now it is simply being restored.

3. I will uniformly change the display to show null as text.

**Review by lishunyang12 — COMMENTED** [review]
## Review of PR #2839

Overall this PR addresses a real CI pain point (nested dicts crashing openpyxl, missing result files breaking downstream pipelines). The approach is reasonable. A few issues to address:

### Issues

**1. Full-width colon in error message (`run_diffusion_benchmark.py`)**

```python
print(f"ERROR：Benchmark script exited with code {process.returncode}")
```

The character after `ERROR` is a full-width colon `：` (U+FF1A) instead of an ASCII colon `:`. This is almost certainly unintentional and will look odd in CI logs. Please fix to use a normal colon.

**2. Silent failure in `run_diffusion_benchmark.py` — original `raise RuntimeError` removed**

The original code raised `RuntimeError` when the benchmark subprocess failed (non-zero exit code). The PR replaces this with a `print()` and falls through to the template fallback. This means `run_benchmark()` will now *return successfully* even when the benchmark subprocess crashes. The caller (`test_benchmark` at the pytest level) will then see a "passing" test with all-zero metrics.

This is intentional for the fallback goal, but it means **test failures are now silently swallowed**. The pytest test case will pass even though the actual benchmark failed. Consider at minimum logging at `WARNING` or `ERROR` level (using `logging.warning(...)`) instead of `print()`, so these are easier to find in CI logs. Even better, consider setting a flag on the returned record (e.g., `"fallback": true`) so downstream consumers (Excel/HTML reports, dashboards) can distinguish real results from fallback placeholders.

**3. Overly complex template extraction in `run_diffusion_benchmark.py`**

The fallback logic in `run_diffusion_benchmark.py` does this:

```python
with open(DIFFUSION_RESULT_TEMPLATE_PATH, encoding="utf-8") as f:
    template_payload = json.load(f)
template_metrics: dict[str, Any] = {}
if isinstance(template_payload, list) and template_payload:
    first_item = template_payload[0]
    if isinstance(first_item, dict):
        result_block = first_item.get("result")
        ...
```

You control the template file — its structure is known and fixed. This defensive multi-branch parsing is unnecessary complexity. Either (a) make the template file directly contain just the metrics dict that's needed here, or (b) simplify to a single known path like `template_payload[0]["result"]` with a clear comment that the template structure is fixed.

**4. The `run_benchmark.py` fallback writes the full template but downstream reads it back immediately**

```python
if not os.path.exists(result_path):
    with open(OMNI_RESULT_TEMPLATE_PATH, encoding="utf-8") as f:
        template_result = json.load(f)
    ...
    with open(result_path, "w", encoding="utf-8") as f:
        json.dump(template_result, f, ...)
with open(result_path, encoding="utf-8") as f:
    result = json.load(f)
```

This writes the template to disk and then immediately reads it back. You could simplify by assigning `result = template_result` directly and skipping the redundant disk round-trip. The file still needs to be written for artifact upload, but you don't need to re-read it.

### Minor / Nits

- The new accuracy test step in the CI YAML (`Diffusion X2I(&A&T) · Accuracy Test`) seems unrelated to the bug fix or fallback mechanism. Consider splitting it into a separate PR for cleaner history, or at least mention it in the PR description.
- Template JSON files look correct and match the expected schemas. Good.
- The `_to_excel_compatible()` function in `generate_nightly_perf_excel.py` is clean and handles edge cases well.
- Unit tests are solid and actually exercise the generation pipeline end-to-end with template data.
- CI YAML: good fix adding `EXIT4` to the bitwise-OR exit code (`exit $$((EXIT1 | EXIT2 | EXIT3 | EXIT4))`), which was missing before.

**Comment by yenuo26** [comment]
> ## Review of PR #2839
> Overall this PR addresses a real CI pain point (nested dicts crashing openpyxl, missing result files breaking downstream pipelines). The approach is reasonable. A few issues to address:
> 
> ### Issues
> **1. Full-width colon in error message (`run_diffusion_benchmark.py`)**
> 
> ```python
> print(f"ERROR：Benchmark script exited with code {process.returncode}")
> ```
> 
> The character after `ERROR` is a full-width colon `：` (U+FF1A) instead of an ASCII colon `:`. This is almost certainly unintentional and will look odd in CI logs. Please fix to use a normal colon.
> 
> **2. Silent failure in `run_diffusion_benchmark.py` — original `raise RuntimeError` removed**
> 
> The original code raised `RuntimeError` when the benchmark subprocess failed (non-zero exit code). The PR replaces this with a `print()` and falls through to the template fallback. This means `run_benchmark()` will now _return successfully_ even when the benchmark subprocess crashes. The caller (`test_benchmark` at the pytest level) will then see a "passing" test with all-zero metrics.
> 
> This is intentional for the fallback goal, but it means **test failures are now silently swallowed**. The pytest test case will pass even though the actual benchmark failed. Consider at minimum logging at `WARNING` or `ERROR` level (using `logging.warning(...)`) instead of `print()`, so these are easier to find in CI logs. Even better, consider setting a flag on the returned record (e.g., `"fallback": true`) so downstream consumers (Excel/HTML reports, dashboards) can distinguish real results from fallback placeholders.
> 
> **3. Overly complex template extraction in `run_diffusion_benchmark.py`**
> 
> The fallback logic in `run_diffusion_benchmark.py` does this:
> 
> ```python
> with open(DIFFUSION_RESULT_TEMPLATE_PATH, encoding="utf-8") as f:
>     template_payload = json.load(f)
> template_metrics: dict[str, Any] = {}
> if isinstance(template_payload, list) and template_payload:
>     first_item = template_payload[0]
>     if isinstance(first_item, dict):
>         result_block = first_item.get("result")
>         ...
> ```
> 
> You control the template file — its structure is known and fixed. This defensive multi-branch parsing is unnecessary complexity. Either (a) make the template file directly contain just the metrics dict that's needed here, or (b) simplify to a single known path like `template_payload[0]["result"]` with a clear comment that the template structure is fixed.
> 
> **4. The `run_benchmark.py` fallback writes the full template but downstream reads it back immediately**
> 
> ```python
> if not os.path.exists(result_path):
>     with open(OMNI_RESULT_TEMPLATE_PATH, encoding="utf-8") as f:
>         template_result = json.load(f)
>     ...
>     with open(result_path, "w", encoding="utf-8") as f:
>         json.dump(template_result, f, ...)
> with open(result_path, encoding="utf-8") as f:
>     result = json.load(f)
> ```
> 
> This writes the template to disk and then immediately reads it back. You could simplify by assigning `result = template_result` directly and skipping the redundant disk round-trip. The file still needs to be written for artifact upload, but you don't need to re-read it.
> 
> ### Minor / Nits
> * The new accuracy test step in the CI YAML (`Diffusion X2I(&A&T) · Accuracy Test`) seems unrelated to the bug fix or fallback mechanism. Consider splitting it into a separate PR for cleaner history, or at least mention it in the PR description.
> * Template JSON files look correct and match the expected schemas. Good.
> * The `_to_excel_compatible()` function in `generate_nightly_perf_excel.py` is clean and handles edge cases well.
> * Unit tests are solid and actually exercise the generation pipeline end-to-end with template data.
> * CI YAML: good fix adding `EXIT4` to the bitwise-OR exit code (`exit $$((EXIT1 | EXIT2 | EXIT3 | EXIT4))`), which was missing before.

1,3,4 fixed
2 Even if an exception is not thrown here, the assert will still fail, causing the pytest to fail.

File: .buildkite/test-nightly.yml [hunk 1]
```diff
@@ -110,8 +110,12 @@ steps:
         timeout_in_minutes: 180
         commands:
           - export BENCHMARK_DIR=tests/dfx/perf/results
-          - pytest -s -v tests/dfx/perf/scripts/run_benchmark.py --test-config-file tests/dfx/perf/tests/test_qwen_omni.json
-          - buildkite-agent artifact upload "tests/dfx/perf/results/*.json"
+          - |
+            set +e
+            pytest -s -v tests/dfx/perf/scripts/run_benchmark.py --test-config-file tests/dfx/perf/tests/test_qwen_omni.json
+            EXIT=$$?
+            buildkite-agent artifact upload "tests/dfx/perf/results/*.json" || true
```
Comment (fhfuih) [inline .buildkite/test-nightly.yml]: `set +e` means to continue running even with errors. So the `|| true` here seems to be useless. (Is it? Not 100% sure about bash script 😁 )

(And there are two more places writing like so)

Comment (yenuo26) [reply .buildkite/test-nightly.yml]: i think you are right, i will fix it

**Comment by yenuo26** [comment]
nightly success, the failure is due to two known issues: https://github.com/vllm-project/vllm-omni/issues/2863, https://github.com/vllm-project/vllm-omni/issues/2862
fd8b8031-9837-4bb7-9ba1-9f3b34514cca

**Review by Gaohan123 — APPROVED** [review]
LGTM. Thanks
